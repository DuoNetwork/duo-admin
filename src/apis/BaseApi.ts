import { IBaseMarketData, ITrade } from '@finbook/duo-market-data';
import ws from 'ws';
import * as CST from '../common/constants';
import { ISource, ISourceAsset, ISourceSettings } from '../common/types';
import sources from '../static/sources.json';
import dbUtil from '../utils/dbUtil';
import util from '../utils/util';

export default abstract class BaseApi {
	public source: string = 'source';

	public settings: ISourceSettings = {
		priceInversed: false,
		quoteInversed: false,
		separator: 'separator',
		isLowercase: false,
		tradesInterval: CST.API_FETCH_TRADE_INTERVAL,
		supportWS: false,
		filterByTimestamp: false,
		wsLink: ''
	};

	public isInitialized: boolean = false;
	public assetsInfo: { [code: string]: ISourceAsset } = {};
	public sourcePairMapping: { [srcPair: string]: string } = {};
	public last: { [localPair: string]: string } = {};
	public tradeStatusLastUpdatedAt: { [key: string]: number } = {};

	public init() {
		if (this.isInitialized) return;

		this.assetsInfo = (sources as { [key: string]: ISource })[this.source].assets;
		this.settings = (sources as { [key: string]: ISource })[this.source].settings;
		util.logInfo(`${this.source} Api initialized`);
	}

	public getSourcePairs(assetIds: string[]): string[] {
		util.logDebug('assets ' + assetIds.join('.'));
		const sourcePairs = [];
		const tickers = Object.keys(this.assetsInfo)
			.map(ticker => (assetIds.includes(this.assetsInfo[ticker].mapping) ? ticker : ''))
			.filter(ticker => ticker !== '');

		for (const baseTicker of tickers)
			for (const quoteTicker of tickers)
				if (
					quoteTicker !== baseTicker &&
					Object.keys(this.assetsInfo[baseTicker].quote).includes(quoteTicker)
				) {
					const quoteLocalCode = this.assetsInfo[quoteTicker].mapping;
					const baseLocalCode = this.assetsInfo[baseTicker].mapping;
					const localPair = quoteLocalCode + '|' + baseLocalCode;

					let sourcePair = '';
					if (this.settings.quoteInversed)
						sourcePair = this.settings.isLowercase
							? baseTicker.toLowerCase() +
							this.settings.separator +
							quoteTicker.toLowerCase()
							: baseTicker.toUpperCase() +
							this.settings.separator +
							quoteTicker.toUpperCase();
					else
						sourcePair = this.settings.isLowercase
							? quoteTicker.toLowerCase() +
							this.settings.separator +
							baseTicker.toLowerCase()
							: quoteTicker.toUpperCase() +
							this.settings.separator +
							baseTicker.toUpperCase();

					this.sourcePairMapping[sourcePair] = localPair;
					sourcePairs.push(sourcePair);
				}

		util.logDebug(
			`BaseApi.getSourcePairs(${JSON.stringify(assetIds)}) => ${JSON.stringify(sourcePairs)}`
		);
		return sourcePairs;
	}

	private filterSortTrades(localPair: string, trades: ITrade[], filterTrades: boolean) {
		//filter and sort in descending order
		const since = filterTrades ? this.last[localPair] : '';

		return trades
			.filter(trade =>
				this.settings.filterByTimestamp
					? trade.timestamp > Number(since)
					: trade.id.localeCompare(since) > 0
			)
			.sort((t1, t2) =>
				this.settings.filterByTimestamp
					? -t1.timestamp + t2.timestamp
					: -t1.id.localeCompare(t2.id)
			);
	}

	public async addTrades(
		localPair: string,
		parsedTrades: ITrade[],
		filterTrades: boolean = true
	) {
		const logMsg: string = `${localPair.split('|')[0]}|${localPair.split('|')[1]}: Received ${
			parsedTrades.length
		} trades; `;

		if (!this.last[localPair]) this.last[localPair] = '';
		if (!this.tradeStatusLastUpdatedAt[localPair]) this.tradeStatusLastUpdatedAt[localPair] = 0;

		const newTrades: ITrade[] = this.filterSortTrades(localPair, parsedTrades, filterTrades);
		const length: number = newTrades.length;

		if (!length) util.logInfo(logMsg + `No new trades since ${this.last[localPair]}`);
		else {
			util.logInfo(
				logMsg +
					`${length} new trades ` +
					(!filterTrades ? '' : `since ${this.last[localPair]}`)
			);

			this.last[localPair] = this.settings.filterByTimestamp
				? newTrades[0].timestamp + ''
				: newTrades[0].id;

			for (let i = 0; i < length; i++) {
				let updateStatusLast: boolean = true;

				//In WS mode, only update trade status.last every 10s
				if (this.settings.supportWS) {
					if (
						util.getUTCNowTimestamp() - this.tradeStatusLastUpdatedAt[localPair] <=
						CST.TRADES_STATUS_LAST_UPDATE_INTERVAL_WS * 1000
					)
						updateStatusLast = false;
				} else this.tradeStatusLastUpdatedAt[localPair] = util.getUTCNowTimestamp();

				await dbUtil.insertTradeData(newTrades[i], !i && updateStatusLast);
			}
		}
	}

	protected parseMarketData(sourcePair: string): IBaseMarketData {
		const [quote, base] = this.sourcePairMapping[sourcePair].split('|');
		return {
			base: base,
			quote: quote,
			source: this.source,
			timestamp: util.getUTCNowTimestamp()
		};
	}

	public async fetchTrades(sourcePairs: string[]) {
		if (!this.settings.supportWS) {
			util.logInfo('Using REST');
			for (const sourcePair of sourcePairs) {
				await this.fetchTradesREST(sourcePair);
				global.setInterval(
					() => this.fetchTradesREST(sourcePair),
					this.settings.tradesInterval * 1000
				);
			}
		} else {
			util.logInfo('Using WS');
			this.fetchTradesWS(sourcePairs);
		}
	}

	public handleWSTradeMessage(msg: string, extraParams: any = {}): void {
		throw new Error(msg + extraParams.toString());
	}

	public handleWSTradeOpen(sourcePairs: string[], w: ws) {
		throw new Error('' + sourcePairs.join(',') + w.url);
	}

	public fetchTradesWS(sourcePairs: string[]): any {
		const w = new ws(this.settings.wsLink);

		w.on('open', () => {
			this.handleWSTradeOpen(sourcePairs, w);
		});

		w.on('message', m => this.handleWSTradeMessage(m.toString(), w));

		w.on('close', (code: number, reason: string) => {
			util.logError('connection closed ' + code + ' ' + reason);
			w.removeAllListeners();
			w.terminate();
			global.setTimeout(() => this.fetchTradesWS(sourcePairs), 1000);
		});

		w.on('error', (error: Error) => {
			util.logError(error);
			w.removeAllListeners();
			w.terminate();
			global.setTimeout(() => this.fetchTradesWS(sourcePairs), 1000);
		});
		return w;
	}

	public abstract fetchTradesREST(sourcePair: string): Promise<void>;
}
