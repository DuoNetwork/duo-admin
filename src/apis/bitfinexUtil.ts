import ws from 'ws';
import sqlUtil from '../sqlUtil';
import * as CST from '../constants';
import { Trade } from '../types';

export class BitfinexUtil {
	parseTrade(trade: object): Trade {
		const amount: number = parseFloat(trade[2]);
		let trade_type: string = 'buy';
		if (amount > 0) {
			trade_type = 'buy';
		} else {
			trade_type = 'sell';
		}
		return {
			source: CST.EXCHANGE_BITFINEX,
			tradeId: trade[0] + '',
			price: trade[3] + '',
			amount: Math.abs(amount) + '',
			tradeType: trade_type,
			sourceTimestamp: trade[1] + ''
		};
	}

	parseApiResponse(msg: string) {
		let parsedJson = JSON.parse(msg.toString());
		// console.log(parsedJson);
		if (parsedJson != undefined) {
			// handle the snapshot
			if (
				parsedJson.event === undefined &&
				parsedJson[1] != 'hb' &&
				!(parsedJson[1] == 'te' || parsedJson[1] == 'tu')
			) {
				// console.log(parsedJson);
				const snapshotArr = parsedJson[1];
				snapshotArr.forEach(trade => {
					const parsedTrade: Trade = this.parseTrade(trade);
					// console.log(parsedTrade);
					sqlUtil.insertSourceData(parsedTrade);
				});
			} else if (parsedJson[1] != 'hb' && parsedJson[1] == 'te') {
				parsedJson = parsedJson[2];

				const parsedTrade: Trade = this.parseTrade(parsedJson);
				sqlUtil.insertSourceData(parsedTrade);
			}
		}
	}

	// Version 2 WebSocket API ---
	fetchTrades() {
		const w = new ws('wss://api.bitfinex.com/ws/2');

		w.on('message', msg => this.parseApiResponse(msg.toString()));

		const msg = JSON.stringify({
			event: 'subscribe',
			channel: 'trades',
			symbol: 'ETHUSD'
		});

		w.on('open', () => {
			console.log('[Bitfinex]-WebSocket is open');
			w.send(msg);
			console.log('subscribe trade');
		});

		w.on('close', () => {
			console.log('[Bitfinex]-WebSocket is close now');
			console.log('close DB');
		});
	}
}
const bitfinexUtil = new BitfinexUtil();
export default bitfinexUtil;
