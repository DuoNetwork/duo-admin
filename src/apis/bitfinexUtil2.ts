import ws from 'ws';
import sqlUtil from '../sqlUtil';
import * as CST from '../constants';

export class BitfinexUtil {
	// Version 2 WebSocket API ---
	parseTrade(trade: object): string[] {
		const amount: number = parseFloat(trade[2]);
		let trade_type: string = 'buy';
		if (amount > 0) {
			trade_type = 'buy';
		} else {
			trade_type = 'sell';
		}
		return [trade[0], trade[3] + '', Math.abs(amount) + '', trade_type, trade[1]];
	}

	fetchETHTradesByOwnWebSocket() {
		const w = new ws('wss://api.bitfinex.com/ws/2');

		w.on('message', msg => {
			let parsedJson = JSON.parse(msg.toString());
			if (parsedJson != undefined) {
				// handle the snopshot
				if (
					parsedJson.event === undefined &&
					parsedJson[1] != 'hb' &&
					!(parsedJson[1] == 'te' || parsedJson[1] == 'tu')
				) {
					// console.log(parsedJson);
					const snopshotArr = parsedJson[1];
					snopshotArr.forEach(trade => {

						let tradeID: string,
							price: string,
							amount: string,
							tradeType: string,
							exchangeTimeStamp: string;

						[tradeID, price, amount, tradeType, exchangeTimeStamp] = this.parseTrade(
							trade
						);

						sqlUtil.insertDataIntoMysql(
							CST.EXCHANGE_BITFINEX,
							tradeID,
							price,
							amount,
							tradeType,
							exchangeTimeStamp
						);
					});
				} else if (parsedJson[1] != 'hb' && parsedJson[1] == 'te') {
					parsedJson = parsedJson[2];
					const amount: number = parseFloat(parsedJson[2]);
					let trade_type: string = 'buy';
					if (amount > 0) {
						trade_type = 'buy';
					} else {
						trade_type = 'sell';
					}
					sqlUtil.insertDataIntoMysql(
						CST.EXCHANGE_BITFINEX,
						parsedJson[0],
						parsedJson[3] + '',
						Math.abs(amount) + '',
						trade_type,
						parsedJson[1]
					);
				}
			}
		});

		const msg = JSON.stringify({
			event: 'subscribe',
			channel: 'trades',
			symbol: 'ETHUSD'
		});

		w.on('open', () => {
			console.log('[Bitfinex]-WebSocket is open');
			w.send(msg);
			console.log('subscribe trade');
		});

		w.on('close', () => {
			console.log('[Bitfinex]-WebSocket is close now');
			console.log('close DB');
		});
	}
}
const bitfinexUtil = new BitfinexUtil();
export default bitfinexUtil;
