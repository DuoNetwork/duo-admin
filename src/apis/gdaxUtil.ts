import sqlUtil from '../sqlUtil';
import util from '../util';
import * as CST from '../constants';
import { Trade } from '../types';

const INTERVAL_SECS = 2;

export class GdaxUtil {

	parseTrade(trade: {[key: string]: string}): Trade {
		return {
			source: CST.EXCHANGE_GDAX,
			id: trade.trade_id,
			price: Number(trade.price),
			amount: Math.abs(Number(trade.size)),
			timestamp: new Date(trade.time).valueOf()
		};
	}

	async fetchETHTradesByRestfulAPI() {
		const data = await util.get('https://api.gdax.com:443/products/ETH-USD/trades');
		const parsedData: Array<{ [key: string]: string }> = JSON.parse(data);

		parsedData.forEach(item => {
			// console.log(item);
			sqlUtil.insertSourceData(
				this.parseTrade(item)
			);
		});
	}

	startFetching() {
		setInterval(() => this.fetchETHTradesByRestfulAPI(), INTERVAL_SECS * 1000);
	}
}

const gdaxUtil = new GdaxUtil();
export default gdaxUtil;
