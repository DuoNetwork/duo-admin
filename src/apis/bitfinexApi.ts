import { Constants, ITrade } from '@finbook/duo-market-data';
import * as CST from '../common/constants';
import {IBitfinexRawTradeRest, IBitfinexRawTradeWS } from '../common/types';
import httpUtil from '../utils/httpUtil';
import util from '../utils/util';
import BaseApi from './BaseApi';

const BFX = require('bitfinex-api-node');

export class BitfinexApi extends BaseApi {
	public source: string = Constants.API_BITFINEX;

	protected parseTradeWS(sourcePair: string, trade: IBitfinexRawTradeWS): ITrade {
		const { base, quote } = this.parseMarketData(sourcePair);
		return {
			base: base,
			quote: quote,
			source: this.source,
			id: trade.id + '',
			price: Number(trade.price),
			amount: Math.abs(Number(trade.amount)),
			timestamp: Number(trade.timestamp) * 1000
		};
	}

	protected parseTradeREST(sourcePair: string, trade: IBitfinexRawTradeRest): ITrade {
		const { base, quote } = this.parseMarketData(sourcePair);
		return {
			base: base,
			quote: quote,
			source: this.source,
			id: trade.tid + '',
			price: Number(trade.price),
			amount: Number(trade.amount),
			timestamp: trade.timestamp * 1000
		};
	}

	public async handleWSTradeMessage(m: string, sourcePair: string) {
		const data = JSON.parse(m);

		// Will ignore batch of trades, because snapshot does not contain trade id.
		// if single trade and is 'tu' event
		if (!Array.isArray(data) && data.hasOwnProperty('id') && sourcePair)
			await this.addTrades(
				this.sourcePairMapping[sourcePair],
				[this.parseTradeWS(sourcePair, data)],
				false
			);
	}

	public async fetchTradesREST(sourcePair: string) {
		const localPair = this.sourcePairMapping[sourcePair];
		const url: string =
			CST.API_BFX_BASE_URL +
			CST.API_BFX_VERSION +
			CST.API_BFX_TRADE +
			sourcePair +
			(this.last[localPair] ? util.composeQuery({ since: this.last[localPair] }) : '');
		util.logInfo(url);

		const result: IBitfinexRawTradeRest[] = JSON.parse(await httpUtil.get(url));
		await this.addTrades(
			this.sourcePairMapping[sourcePair],
			result.map(trade => this.parseTradeREST(sourcePair, trade))
		);
	}

	public fetchTradesWS(sourcePairs: string[]): any {
		const bfx = new BFX(null, null, {
			version: 1,
			transform: true
		});

		const w = bfx.ws;
		w.on('open', () => {
			util.logInfo('Connected');
			for (const pair of sourcePairs) {
				w.subscribeTrades(pair);
				util.logInfo(`Subscribed to trades ${pair}`);
			}
		});

		w.on('trade', (symbol: string, trades: any) => {
			this.handleWSTradeMessage(JSON.stringify(trades), symbol);
		});

		w.on('close', (code: number, reason: string) => {
			util.logError('connection closed ' + code + ' ' + reason);
			w.close();
			global.setTimeout(() => this.fetchTradesWS(sourcePairs), 1000);
		});

		w.on('error', (error: Error) => {
			util.logError(error);
			w.close();
			global.setTimeout(() => this.fetchTradesWS(sourcePairs), 1000);
		});
		return w;
	}
}
const bitfinexApi = new BitfinexApi();
export default bitfinexApi;
