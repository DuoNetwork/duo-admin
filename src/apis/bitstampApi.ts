import { Constants, ITrade } from '@finbook/duo-market-data';
import Pusher from 'pusher-js';
import * as CST from '../common/constants';
import { IBitstampRawTradeRest, IBitstampRawTradeWs } from '../common/types';
import httpUtil from '../utils/httpUtil';
import util from '../utils/util';
import BaseApi from './BaseApi';

export class BitfinexApi extends BaseApi {
	public source: string = Constants.API_BITSTAMP;

	protected parseTradeWS(sourcePair: string, trade: IBitstampRawTradeWs): ITrade {
		const { base, quote } = this.parseMarketData(sourcePair);
		return {
			base: base,
			quote: quote,
			source: this.source,
			id: trade.id + '',
			price: trade.price,
			amount: Math.abs(trade.amount),
			timestamp: Number(trade.timestamp) * 1000
		};
	}

	protected parseTradeREST(sourcePair: string, trade: IBitstampRawTradeRest): ITrade {
		const { base, quote } = this.parseMarketData(sourcePair);
		return {
			base: base,
			quote: quote,
			source: this.source,
			id: trade.tid + '',
			price: Number(trade.price),
			amount: Number(trade.amount),
			timestamp: trade.date * 1000
		};
	}

	public async handleWSTradeMessage(m: string, sourcePair: string) {
		const data = JSON.parse(m);
		await this.addTrades(
			this.sourcePairMapping[sourcePair],
			[this.parseTradeWS(sourcePair, data)],
			false
		);
	}

	public async fetchTradesREST(sourcePair: string) {
		const localPair = this.sourcePairMapping[sourcePair];
		const url: string =
			CST.API_BST_BASE_URL +
			CST.API_BST_VERSION +
			CST.API_BST_TRANSACTIONS +
			sourcePair +
			(this.last[localPair] ? util.composeQuery({ since: this.last[localPair] }) : '');
		util.logInfo(url);

		const result: IBitstampRawTradeRest[] = JSON.parse(await httpUtil.get(url));
		await this.addTrades(
			this.sourcePairMapping[sourcePair],
			result.map(trade => this.parseTradeREST(sourcePair, trade))
		);
	}

	public fetchTradesWSForPair(sourcePair: string): any {
		const socket = new Pusher(CST.API_BST_PUSHER_APP_KEY);
		socket.bind('trade', trade => {
			this.handleWSTradeMessage(JSON.stringify(trade), sourcePair);
		});

		util.logInfo(`Subscribed to trades ${sourcePair}`);
		socket.subscribe(`live_trades_${sourcePair}`);
		socket.connection.bind('error', err => {
			util.logError(`connection error, error: ${JSON.stringify(err)}`);
			let timeOutDuration = 1000;
			if (err.error.data.code === 4004) {
				util.logError('Over limit!');
				timeOutDuration = 30000;
			}
			socket.disconnect();
			global.setTimeout(() => this.fetchTradesWSForPair(sourcePair), timeOutDuration);
		});

		socket.connection.bind('closed', () => {
			util.logError('connection closed');
			global.setTimeout(() => this.fetchTradesWSForPair(sourcePair), 1000);
		});
		return socket;
	}

	public fetchTradesWS(sourcePairs: string[]) {
		for (const sourcePair of sourcePairs) this.fetchTradesWSForPair(sourcePair);
	}
}
const bitfinexApi = new BitfinexApi();
export default bitfinexApi;
