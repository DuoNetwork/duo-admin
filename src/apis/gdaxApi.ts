import moment from 'moment';
import * as CST from '../common/constants';
import { ITrade } from '../common/types';
import httpUtil from '../utils/httpUtil';
import util from '../utils/util';
import BaseApi from './BaseApi';

export class GdaxApi extends BaseApi {
	public source: string = CST.API_GDAX;
	public parseTrade(sourcePair: string, trade: { [key: string]: string | number }): ITrade {
		const { base, quote } = this.parseMarketData(sourcePair);
		return {
			base: base,
			quote: quote,
			source: CST.API_GDAX,
			id: trade.trade_id + '',
			price: Number(trade.price),
			amount: Math.abs(Number(trade.size)),
			timestamp: moment(trade.time).valueOf()
		};
	}

	public async fetchTradesREST(sourcePair: string) {
		const url: string = CST.API_GDAX_BASE_URL + '/' + sourcePair + CST.API_GDAX_TRADE;
		util.logInfo(url);
		const res = await httpUtil.get(url);
		const result: Array<{ [key: string]: string }> = JSON.parse(res);
		await this.addTrades(
			this.sourcePairMapping[sourcePair],
			result.map(trade => this.parseTrade(sourcePair, trade))
		);
	}

	public fetchTradesWS(): void {
		throw new Error('no ws');
	}
}

const gdaxApi = new GdaxApi();
export default gdaxApi;
