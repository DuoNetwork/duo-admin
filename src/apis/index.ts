import * as CST from '../constants';
import BaseApi from './BaseApi';
import bitfinexApi from './bitfinexApi';
import gdaxApi from './gdaxApi';
import geminiApi from './geminiApi';
import krakenApi from './krakenApi';

const apis: { [key: string]: BaseApi } = {
	[CST.EXCHANGE_BITFINEX]: bitfinexApi,
	[CST.EXCHANGE_GEMINI]: geminiApi,
	[CST.EXCHANGE_KRAKEN]: krakenApi,
	[CST.EXCHANGE_GDAX]: gdaxApi
};

export default apis;
