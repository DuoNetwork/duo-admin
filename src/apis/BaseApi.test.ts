import moment from 'moment';
import apis from '../apis';
import * as CST from '../common/constants';
import dbUtil from '../utils/dbUtil';
import util from '../utils/util';

const parsedTrades = [
	{
		id: 'ID_1',
		price: 0.05,
		amount: 1,
		source: 'source',
		base: 'USD',
		quote: 'ETH',
		timestamp: 1234567892
	},
	{
		id: 'ID_2',
		price: 0.05,
		amount: 1,
		source: 'source',
		base: 'USD',
		quote: 'ETH',
		timestamp: 1234567891
	}
];

for (const source in apis) {
	const api = apis[source];
	api.isInitialized = false;
	api.init();

	test(`getSourcePairs ${source} `, async () => {
		if (Object.keys(api.assetsInfo).length)
			api.assetsInfo = {
				baseTicker: {
					base: {},
					quote: { quoteTicker: true },
					mapping: 'base',
					name: 'base name'
				},
				quoteTicker: {
					base: { baseTicker: true },
					quote: {},
					mapping: 'quote',
					name: 'quote name'
				}
			};

		util.getUTCNowTimestamp = jest.fn(() => moment.utc('20180924').valueOf());
		const sourceInstruments = api.getSourcePairs(['quote', 'base']);
		expect(sourceInstruments).toMatchSnapshot();
	});

	//
	test(`addTrades WS ${source} `, async () => {
		api.isInitialized = false;

		api.init();
		if (api.settings.supportWS) {
			util.getUTCNowTimestamp = jest.fn(
				() => CST.TRADES_STATUS_LAST_UPDATE_INTERVAL_WS * 1000 + 1
			);
			dbUtil.insertTradeData = jest.fn(() => Promise.resolve({}));

			const localPair = `${parsedTrades[0].quote}|${parsedTrades[0].base}`;

			await api.addTrades(localPair, parsedTrades, false);
			expect(api.settings.supportWS).toEqual(true);
			expect((dbUtil.insertTradeData as jest.Mock<void>).mock.calls).toMatchSnapshot();
		}
	});

	test(`addTrades REST ${source} - filterBy (${
		api.settings.filterByTimestamp ? 'timestamp > 1234567891' : 'id > ID_1'
	}) `, async () => {
		api.isInitialized = false;
		api.init();
		api.settings.supportWS = false;

		util.getUTCNowTimestamp = jest.fn(() => CST.TRADES_STATUS_LAST_UPDATE_INTERVAL_WS * 1000); // expect to have no effect
		dbUtil.insertTradeData = jest.fn(() => Promise.resolve({}));

		const localPair = `${parsedTrades[0].quote}|${parsedTrades[0].base}`;
		api.last[localPair] = api.settings.filterByTimestamp ? '1234567891' : 'ID_1';

		await api.addTrades(localPair, parsedTrades);
		expect(api.settings.supportWS).toBe(false);
		expect((dbUtil.insertTradeData as jest.Mock<void>).mock.calls).toMatchSnapshot();
	});
}
