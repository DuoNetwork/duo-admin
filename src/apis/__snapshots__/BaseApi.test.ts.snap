// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addTrades WS  1`] = `
Array [
  Array [
    Object {
      "amount": 1,
      "base": "USD",
      "id": "ID_1",
      "price": 0.05,
      "quote": "ETH",
      "source": "source",
      "timestamp": 1234567892,
    },
    true,
  ],
  Array [
    Object {
      "amount": 1,
      "base": "USD",
      "id": "ID_2",
      "price": 0.05,
      "quote": "ETH",
      "source": "source",
      "timestamp": 1234567891,
    },
    false,
  ],
]
`;

exports[`addTrades WS  no trade 1`] = `Array []`;

exports[`addTrades WS no tradeStatusLastUpdatedAt 1`] = `
Array [
  Array [
    Object {
      "amount": 1,
      "base": "USD",
      "id": "ID_1",
      "price": 0.05,
      "quote": "ETH",
      "source": "source",
      "timestamp": 1234567892,
    },
    false,
  ],
  Array [
    Object {
      "amount": 1,
      "base": "USD",
      "id": "ID_2",
      "price": 0.05,
      "quote": "ETH",
      "source": "source",
      "timestamp": 1234567891,
    },
    false,
  ],
]
`;

exports[`already initialized 1`] = `
Object {
  "filterByTimestamp": true,
  "inflationRequired": false,
  "isLowercase": false,
  "orderBooksInterval": 2,
  "priceInversed": false,
  "quoteInversed": true,
  "separator": "",
  "supportWS": true,
  "tradesInterval": 2,
  "wsLink": "",
}
`;

exports[`fetchTrades WS  1`] = `
Array [
  Array [
    "open",
    [Function],
  ],
  Array [
    "message",
    [Function],
  ],
  Array [
    "close",
    [Function],
  ],
  Array [
    "error",
    [Function],
  ],
]
`;

exports[`fetchTrades WS  2`] = `
Array [
  Array [
    "quote-base",
  ],
  Object {
    "on": [MockFunction] {
      "calls": Array [
        Array [
          "open",
          [Function],
        ],
        Array [
          "message",
          [Function],
        ],
        Array [
          "close",
          [Function],
        ],
        Array [
          "error",
          [Function],
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    },
    "removeAllListeners": [MockFunction],
    "terminate": [MockFunction],
  },
]
`;

exports[`fetchTrades WS  3`] = `
Array [
  "message from ws server",
  Object {
    "on": [MockFunction] {
      "calls": Array [
        Array [
          "open",
          [Function],
        ],
        Array [
          "message",
          [Function],
        ],
        Array [
          "close",
          [Function],
        ],
        Array [
          "error",
          [Function],
        ],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    },
    "removeAllListeners": [MockFunction],
    "terminate": [MockFunction],
  },
]
`;

exports[`fetchTrades WS  4`] = `
Array [
  Array [
    [Function],
    1000,
  ],
  Array [
    [Function],
    1000,
  ],
]
`;

exports[`fetchTrades WS  5`] = `
Array [
  Array [
    Array [
      "quote-base",
    ],
  ],
  Array [
    Array [
      "quote-base",
    ],
  ],
]
`;

exports[`fetchTrades, supportWs false 1`] = `
Array [
  Array [
    [Function],
    2000,
  ],
]
`;

exports[`fetchTrades, supportWs false 2`] = `
Array [
  Array [
    "quote-base",
  ],
  Array [
    "quote-base",
  ],
]
`;

exports[`getSourcePairs  1`] = `
Array [
  "BASEQUOTE",
]
`;

exports[`getSourcePairs , not quoteInversed , isLowerCase true 1`] = `
Array [
  "basequote",
]
`;

exports[`getSourcePairs isLowerCase 1`] = `
Array [
  "quotebase",
]
`;

exports[`getSourcePairs quoteInversed 1`] = `
Array [
  "QUOTEBASE",
]
`;

exports[`getSourcePairs, assetId wrong  1`] = `Array []`;

exports[`handleWSTradeMessage 1`] = `"tradeMessage[object Object]"`;

exports[`handleWSTradeMessage, with extra 1`] = `"tradeMessageExtra"`;

exports[`handleWSTradeOpen 1`] = `"pair1,pair2url"`;
